Fixed the window update hangs

davidfowl opened the pull request:
- While computing how many bytes we needed to write, we were returning the remaining number of window bytes (for both the connection and stream windows). Since window updates happen concurrently with the write loop, it was possible to miss and not observe the window update depending on when the update happened in relation to the loop observing this state. We now look at the current state of the connection window and stream window before we decide to queue or wait for a window update under the appropriate locks.


Fixes #41248

davidfowl committed:
Fixed the window update hangs causing tests to fail
- While computing how many bytes we needed to write, we were returning the remainng number of window bytes (for both the connection and stream). Since window updates happen concurrently with the write loop, it was possibe to miss and not observe the window update depending on when the update happened in relation to the loop observing this state. We now look at the current state of the connection window and stream window before we decide to queue or wait for a window update under the appropriate locks.

davidfowl commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -292,6 +260,37 @@ static bool HasStateFlag(Http2OutputProducer.State state, Http2OutputProducer.St
+            // Check the connection window under a lock so that we don't miss window updates
+            if (_connectionWindow == 0)
+            {
+                // We have no more connection window, put this producer in a queue waiting for it to
+                // a window update to resume the connection.
```
```suggestion
                // We have no more connection window, put this producer in a queue waiting for
                // a window update to resume the connection.
```

davidfowl committed:
Apply suggestions from code review

JamesNK commented on src/Grpc/Interop/test/InteropTests/InteropTests.cs:
```diff
@@ -64,7 +64,6 @@ public InteropTests(ITestOutputHelper output)
     public Task UnimplementedMethod() => InteropTestCase("unimplemented_method");
 
     [Fact]
-    [QuarantinedTest("https://github.com/dotnet/aspnetcore/issues/41248")]
```
Policy is to leave it quarantined. Remove attribute after 30 days of passes

davidfowl replied:
Got it

davidfowl replied:
BUT this feels like one that we should un-quarantine since its my change that made it flaky.

Tratcher replied:
Maybe if it were 100% repro before, and had a targeted test to say it was 100% fixed now. Right now we only *think* it's fixed 😁, hence staying quarantined.

davidfowl replied:
It reprod 100% before on my machine.

BrennanConroy replied:
It's just a general rule, but up to devs judgement if they strongly believe they know the cause and the fix.

davidfowl replied:
Yea, I was able to repro locally 100%, debug and root cause the thing.

JamesNK commented:
Is there anyway to unit test this? If it regresses we have some tests that become flaky, but it’s not an obvious break.

davidfowl commented:
This is hard to unit test without adding lots of events that control exactly when things happen.

davidfowl commented on src/Grpc/Interop/test/InteropTests/InteropTests.cs:
```diff
@@ -64,7 +64,6 @@ public InteropTests(ITestOutputHelper output)
     public Task UnimplementedMethod() => InteropTestCase("unimplemented_method");
 
     [Fact]
```
```suggestion
    [Fact]
    [QuarantinedTest("https://github.com/dotnet/aspnetcore/issues/41248")]
```

Tratcher commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -128,10 +128,10 @@ private async Task WriteToOutputPipe()
                     // Check the stream window
-                    var (actual, remainingStream) = producer.ConsumeStreamWindow(buffer.Length);
+                    var actual = producer.ConsumeStreamWindow(buffer.Length);
 
                     // Now check the connection window
-                    (actual, var remainingConnection) = ConsumeConnectionWindow(actual);
+                    actual = ConsumeConnectionWindow(actual);
```
Q: Are we leaking stream window bytes?
Setup: The stream has an available window of 50. The connection has an available window of 40. We try to write 50+ bytes. The steam window is consumed. The connection window is consumed, but then the buffer is sliced to 40 bytes. Are those 10 extra bytes credited back to the stream window for future writes?

davidfowl replied:
Good catch!

davidfowl replied:
Fixed. Please review.

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -292,6 +260,37 @@ static bool HasStateFlag(Http2OutputProducer.State state, Http2OutputProducer.St
+        lock (_windowUpdateLock)
+        {
+            // Check the connection window under a lock so that we don't miss window updates
+            if (_connectionWindow == 0)
```
I think only the stream window can go negative. That's because `SETTINGS_INITIAL_WINDOW_SIZE` can be decreased after the connection starts, but that only affects stream-level flow control. The connection window can only be updated with WINDOW_UPDATES which cannot be negative.

davidfowl replied:
Right!

davidfowl committed:
Check both the stream and connection windows
- Added a test to verify we don't lost bytes if the stream window is more than the connection window.

davidfowl committed:
Can't be negative!

halter73 commented on src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2ConnectionTests.cs:
```diff
@@ -1936,6 +1936,58 @@ public async Task DATA_Sent_DespiteStreamOutputFlowControl_IfEmptyAndEndsStream(
+        await InitializeConnectionAsync(async context =>
+        {
+            var data = new byte[16384 * 4];
+            data.AsSpan().Fill(1);
```
Why are we filling the array with 1s? Doesn't `new byte[16384 * 4]` fit just inside the 64KiB connection window? Apparently it's one byte too large given the test, but I'm trying to understand why.

davidfowl replied:
https://github.com/dotnet/aspnetcore/blob/46554c9183e7ebdda173edd03038f19ea7ac75d3/src/Shared/ServerInfrastructure/Http2/Http2PeerSettings.cs#L14

halter73 replied:
Right, it's *almost* 64KiB.

halter73 commented on src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2ConnectionTests.cs:
```diff
@@ -1936,6 +1936,58 @@ public async Task DATA_Sent_DespiteStreamOutputFlowControl_IfEmptyAndEndsStream(
+            withStreamId: 1).DefaultTimeout();
+
+        // Send more than enough bytes for the window update
+        await SendWindowUpdateAsync(0, (int)Http2PeerSettings.DefaultInitialWindowSize);
```
This works, right?

```suggestion
        await SendWindowUpdateAsync(0, 1);
```

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -855,13 +858,19 @@ private ValueTask<FlushResult> TimeFlushUnsynchronizedAsync()
+    {
+        lock (_windowUpdateLock)
+        {
+            return Math.Min(bytes, _connectionWindow);
```
You could still consume the connection window here and avoid retaking the lock, right?

davidfowl replied:
I rather not update this state outside of the lock ever 😄. Maybe in preview5.

halter73 replied:
I'm suggesting just going back to the old version of ConsumeConnectionWindow more or less since you always consume what's returned from here (since you've already checked the stream window).

```
private long ConsumeConnectionWindow(long bytes)
{
    lock (_windowUpdateLock)
    {
        var actual = Math.Min(bytes, _connectionWindow);
        _connectionWindow -= actual;
        return actual;
    }
}

davidfowl replied:
I see, initially I had this but I decided to keep both consistent.

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -233,42 +237,10 @@ static bool HasStateFlag(Http2OutputProducer.State state, Http2OutputProducer.St
                     // When the window update is sent, the producer will be re-queued if needed.
                     else if (hasMoreData && !aborted && !waitingForWindowUpdates)
```
Can we simplify this by removing `waitingForWindowUpdates` since `TryScheduleNextWrite` will check this independently anyway? `waitingForWindowUpdates` only applies to stream window updates, right?

davidfowl replied:
waitingForWindowUpdates applies to both connection and stream window currently.

davidfowl committed:
PR feedback
